Docker: containers

Durf1985 opened the pull request:
## In the process done:

* Creating own custom image
* Creating image for terraform with using packer and ansible role
* Running a docker container on a virtual machine running Terraform in GCP

## How to setup the project:

* In directory `~/your_repo/docker-monolith/infra/packer` execute follow command

```bash
packer build -var-file=variables.json app.json
```

* In `~/your_repo/docker-monolith/infra/terraform` directory, run the 

```bash
terraform apply
```

* In `~/your_repo/docker-monolith/infra/ansible` directory run

```bash
ansible-playbook playbooks/monolith_container_launch.yml
```

## How to check the project:

-Follow the link <http://external_ip_your_vm:9292>

## PR checklist

The label with the homework number, in label description - homework lecture and topic

Durf1985 committed:
DO-107 Create and run monolith on docker-machine

Created a set of configuration files to create a docker-image

loktionovam commented on docker-monolith/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+FROM ubuntu:18.04
+RUN apt update
+RUN apt-get install -y software-properties-common wget sudo git
+RUN apt-get clear
```
This doesn't work

```
 ---> Running in df695d215b34
E: Invalid operation clear
The command '/bin/sh -c apt-get clear' returned a non-zero code: 100
```

remove it

loktionovam commented on docker-monolith/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+RUN apt update
+RUN apt-get install -y software-properties-common wget sudo git
+RUN apt-get clear
+RUN apt -y upgrade
```
Do not use `upgrade` inside the dockerfile because your image become unpredictable

loktionovam commented on docker-monolith/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+RUN wget -qO - https://www.mongodb.org/static/pgp/server-3.2.asc | sudo apt-key add -
+RUN echo "deb http://repo.mongodb.org/apt/ubuntu xenial/mongodb-org/3.2 multiverse" | sudo tee /etc/apt/sources.list.d/mongodb-org-3.2.list
+RUN apt-get update
+RUN apt-get -y upgrade
```
Do not use `upgrade` inside the dockerfile because your image become unpredictable and this will reset your build cache

Durf1985 committed:
DO-109 Docker monolith infra

A terraform configuration has been created for use with docker

The docker_install.yml playbook and its nested roles were created to
create an image using a packer that is used in the terraform configuration

A monolith_container_launch.yml playbook has been created that runs
the container on an infrastructure already created and prepared with
the help of terraform and packer

loktionovam commented on docker-monolith/infra/ansible/roles/docker_image/defaults/main.yml:
```diff
@@ -0,0 +1,3 @@
+---
+env: local
+os_version: bionic
```
Use garther_facts here

loktionovam commented on docker-monolith/infra/ansible/roles/docker_image/output.txt:
```diff
@@ -0,0 +1,160 @@
+[1;31mWARNING: ansible-lint is no longer tested under Python [0m[1;36m3.8[0m[1;31m and will soon require [0m[1;36m3.9[0m[1;31m. Do not report bugs for this version.[0m
```
This is an unnecessary file

loktionovam commented on docker-monolith/infra/ansible/roles/docker_image/tasks/docker:
```diff
@@ -0,0 +1,42 @@
+---
+- name: Create keyrings directory
```
This is an unnecessary file

loktionovam commented on docker-monolith/infra/ansible/roles/docker_image/tasks/docker.yml:
```diff
@@ -0,0 +1,30 @@
+
+- name: Add Docker repository
+  ansible.builtin.apt_repository:
+    repo: deb [arch=amd64] https://download.docker.com/linux/ubuntu {{ os_version }} stable
```
Use gather_facts here

loktionovam commented on docker-monolith/infra/ansible/roles/monolith/.travis.yml:
```diff
@@ -0,0 +1,29 @@
+---
+language: python
```
I don't use travis so can remove this file

loktionovam commented on docker-monolith/infra/ansible/roles/monolith/README.md:
```diff
@@ -0,0 +1,38 @@
+Role Name
```
It is good idea to write a short README about your role

loktionovam commented on docker-monolith/infra/ansible/roles/monolith/tasks/container_launcher.yml:
```diff
@@ -0,0 +1,25 @@
+- name: Launch a Docker container
+  ansible.builtin.docker_container:
+    name: reddit
+    image: fallgeratoor/otus-reddit:1.0
```
at least it is better to parametrize version

loktionovam commented on docker-monolith/infra/terraform/main.tf:
```diff
@@ -0,0 +1,49 @@
+}
+
+provider "google" {
+  project = var.project //+
```
Unnecessary comments

loktionovam commented on docker-monolith/infra/ansible/roles/monolith/tasks/container_launcher.yml:
```diff
@@ -0,0 +1,25 @@
+    restart_policy: always
+    network_mode: bridge
+    ports:
+      - "9292:9292"
```
In terraform you parametrize port but here not, so you should to parametrize `{{role_name_reddit_monolith_port}}:9292`

loktionovam reviewed (commented):
I suggest you to rename your roles:
- monolith -> reddit_monolith (because `monolith` word is too generic)
- docker_image -> docker_host (because naming is confusing and you do not use docker images at your role at all)

loktionovam commented on README.md:
````diff
@@ -0,0 +1,37 @@
+* To create image in directory `~/your_repo/docker-monolith` execute next command
+  
+```bash
+docker build -t your_own_tag
````
You forgot the `.` at the end of the command

loktionovam commented on docker-monolith/infra/ansible/roles/docker_host/tests/test.yml:
```diff
@@ -0,0 +1,5 @@
+- hosts: localhost
+  remote_user: root
+  roles:
+    - docker_image
```
docker_host should be here

loktionovam commented on docker-monolith/infra/ansible/roles/reddit_monolith/defaults/main.yml:
```diff
@@ -0,0 +1,4 @@
+---
+env: local
+docker_image: fallgeratoor/otus-reddit:1.0
```
it is better to use `reddit_monolith` prefix because `docker_image` is too generic and can be accidentally redefined somewhere else
